src/cryptography/x509.py

```diff
@@ -29,6 +31,147 @@ def __init__(self, msg, parsed_version):
         self.parsed_version = parsed_version
 
 
+class UnknownAttribute(Exception):
+    pass
+
+
+class Attribute(object):
+    def __init__(self, oid, value):
+        if not isinstance(oid, ObjectIdentifier):
+            raise TypeError("oid argument must be an ObjectIdentifier object")
+
+        self._oid = oid
+        self._value = value
+
+    oid = utils.read_only_property("_oid")
+    value = utils.read_only_property("_value")
+
+    def __eq__(self, other):
+        if not isinstance(other, Attribute):
+            return NotImplemented
+
+        return (
+            self.oid == other.oid and
+            self.value == other.value
+        )
+
+    def __ne__(self, other):
+        return not self == other
+
+
+class ObjectIdentifier(object):
+    def __init__(self, oid, name):
+        self._value = oid
+        self._name = name
```

Comment:
And (optionally) some public constants for known oids?

On Tue, Jan 27, 2015 at 9:56 PM, Alex Gaynor <alex.gaynor@gmail.com> wrote:

> How about an intermediary state:
>
> * `name` isn't part of the public API for an OID.
> * `repr()` of an OID contains the name if it's known
> * Constructor for OIDs isn't public so we can punt on this :-)
>
> On Tue, Jan 27, 2015 at 9:50 PM, Paul Kehrer <notifications@github.com>
> wrote:
>
>> In src/cryptography/x509.py
>> <https://github.com/pyca/cryptography/pull/1612#discussion_r23662579>:
>>
>> > +        if not isinstance(other, Attribute):
>> > +            return NotImplemented
>> > +
>> > +        return (
>> > +            self.oid == other.oid and
>> > +            self.value == other.value
>> > +        )
>> > +
>> > +    def __ne__(self, other):
>> > +        return not self == other
>> > +
>> > +
>> > +class ObjectIdentifier(object):
>> > +    def __init__(self, oid, name):
>> > +        self._value = oid
>> > +        self._name = name
>>
>> If we choose to do that we'd need to handle these cases:
>>
>>    - Known OID (name not provided)
>>    - Known OID (name provided and matches)
>>    - Known OID (name provided and does not match)
>>    - Unknown OID (name provided)
>>    - Unknown OID (name not provided)
>>
>> Any others/
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/pyca/cryptography/pull/1612/files#r23662579>.
>>
>
>
>
> --
> "I disapprove of what you say, but I will defend to the death your right
> to say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
> "The people's good is the highest law." -- Cicero
> GPG Key fingerprint: 125F 5C67 DFE9 4084
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084